commit 80099c878a2d60ead7cb7bbbfab89e1eb5782dad
Author: David Faure <faure@kde.org>
Date:   Fri Jun 7 15:03:30 2013 +0200

    Fix misuse of KUrl::relativeUrl, which returns relative urls (duh).
    
    Not paths. So this was leading to double-escaping, especially after fixing
    KUrl::relativeUrl to actually escape as it's supposed to do.
    
    BUG: 317712
    FIXED-IN: 4.10.5

diff --git a/kfile/kfilewidget.cpp b/kfile/kfilewidget.cpp
index 05d158b..fffe753 100644
--- a/kfile/kfilewidget.cpp
+++ b/kfile/kfilewidget.cpp
@@ -1229,6 +1229,24 @@ void KFileWidgetPrivate::setLocationText(const KUrl& url)
     }
 }
 
+static bool isRelativeUrl(const KUrl &baseUrl, const KUrl& url)
+{
+    return KUrl::relativeUrl(baseUrl, url) != url.url();
+}
+
+static QString relativePathOrUrl(const KUrl &baseUrl, const KUrl& url)
+{
+    if (isRelativeUrl(baseUrl, url)) {
+        QString relPath = KUrl::relativePath(baseUrl.path(), url.path());
+        if (relPath.startsWith("./")) {
+            relPath = relPath.mid(2);
+        }
+        return relPath;
+    } else {
+        return url.prettyUrl();
+    }
+}
+
 void KFileWidgetPrivate::setLocationText( const KUrl::List& urlList )
 {
     const KUrl currUrl = ops->url();
@@ -1236,14 +1254,14 @@ void KFileWidgetPrivate::setLocationText( const KUrl::List& urlList )
     if ( urlList.count() > 1 ) {
         QString urls;
         foreach (const KUrl &url, urlList) {
-            urls += QString( "\"%1\"" ).arg( KUrl::relativeUrl(currUrl, url) ) + ' ';
+            urls += QString("\"%1\"").arg(relativePathOrUrl(currUrl, url));
         }
         urls = urls.left( urls.size() - 1 );
 
         setDummyHistoryEntry( urls, QPixmap(), false );
-    } else if ( urlList.count() ) {
+    } else if ( urlList.count() == 1 ) {
         const QPixmap mimeTypeIcon = KIconLoader::global()->loadMimeTypeIcon( KMimeType::iconNameForUrl( urlList[0] ),  KIconLoader::Small );
-        setDummyHistoryEntry( KUrl::relativeUrl(currUrl, urlList[0]), mimeTypeIcon );
+        setDummyHistoryEntry( relativePathOrUrl(currUrl, urlList[0]), mimeTypeIcon );
     } else {
         removeDummyHistoryEntry();
     }
